src/coreclr/vm/interoplibinterface.cpp

```diff
@@ -300,64 +303,49 @@ namespace
+            CQuickArrayList<ExternalObjectContext*> localList;
 
-            // Populate the array
+            // Determine objects to return
             {
                 LockHolder lock(this);
-                Iterator curr = _hashMap.Begin();
                 Iterator end = _hashMap.End();
-
-                ExternalObjectContext* inst;
-                for (objCount = 0; curr != end && objCount < objCountMax; objCount++, curr++)
+                for (Iterator curr = _hashMap.Begin(); curr != end; ++curr)
                 {
-                    inst = *curr;
+                    ExternalObjectContext* inst = *curr;
 
                     // Only add objects that are in the correct thread
                     // context and have the appropriate flags set.
                     if (inst->ThreadContext == threadContext
                         && (withFlags == ExternalObjectContext::Flags_None || inst->IsSet(withFlags)))
                     {
-                        // Separate the wrapper from the tracker runtime prior to
-                        // passing this onto the caller. This call is okay to make
-                        // even if the instance isn't from the tracker runtime.
-                        InteropLib::Com::SeparateWrapperFromTrackerRuntime(inst);
-                        gc.arrRef->SetAt(objCount, inst->GetObjectRef());
+                        localList.Push(inst);
                         STRESS_LOG1(LF_INTEROP, LL_INFO100, "Add EOC to Enumerable: 0x%p\n", inst);
                     }
                 }
             }
 
-            // Make the array the correct size
-            if (objCount < objCountMax)
-            {
-                gc.arrRefTmp = (PTRARRAYREF)AllocateObjectArray(objCount, g_pObjectClass);
-
-                SIZE_T elementSize = gc.arrRef->GetComponentSize();
-
-                void *src = gc.arrRef->GetDataPtr();
-                void *dest = gc.arrRefTmp->GetDataPtr();
+            // Allocate enumerable type to return.
+            gc.arrRef = (PTRARRAYREF)AllocateObjectArray((DWORD)localList.Size(), g_pObjectClass);
 
-                _ASSERTE(sizeof(Object*) == elementSize && "Assumption invalidated in memmoveGCRefs() usage");
-                memmoveGCRefs(dest, src, objCount * elementSize);
-                gc.arrRef = gc.arrRefTmp;
+            // Separate wrappers and insert object into enumerable.
+            for (SIZE_T i = 0; i < localList.Size(); i++)
+            {
+                ExternalObjectContext* inst = localList[i];
+                {
+                    // Separate the wrapper from the tracker runtime prior to
+                    // passing this onto the caller. This call is okay to make
+                    // even if the instance isn't from the tracker runtime.
+                    GCX_PREEMP();
+                    InteropLib::Com::SeparateWrapperFromTrackerRuntime(inst);
+                }
+                gc.arrRef->SetAt(i, inst->GetObjectRef());
```

Comment:
This could drop a null in here, as its a ref handle, and a GC could have occurred with the GCX_PREEMP() above.

Reply:
Removed.